clang/lib/Sema/SemaDecl.cpp

```diff
@@ -17563,6 +17573,16 @@ Sema::ActOnTag(Scope *S, unsigned TagSpec, TagUseKind TUK, SourceLocation KWLoc,
                                           UPPC_FixedUnderlyingType))
         EnumUnderlying = Context.IntTy.getTypePtr();
 
+      // If the underlying type is atomic, we need to adjust the type before
+      // continuing. This only happens in the case we stored a TypeSourceInfo
+      // into EnumUnderlying because the other cases are error recovery up to
+      // this point. But because it's not possible to gin up a TypeSourceInfo
+      // for a non-atomic type from an atomic one, we'll store into the Type
+      // field instead.
+      if (TypeSourceInfo *TI = dyn_cast<TypeSourceInfo *>(EnumUnderlying);
+          TI && TI->getType()->isAtomicType())
+        EnumUnderlying = TI->getType().getAtomicUnqualifiedType().getTypePtr();
```

Comment:
Should we be stripping ALL cv qualifiers here?  Not just when it is atomic?

Reply:
`getAtomicUnqualifiedType()` strips all qualifiers *and* `_Atomic` (which is not a qualifier despite being listed as a qualifier in the standard. Yes, this is terrible.)

Reply:
Right, i just mean "why the branch, why not do the stripping unconditionally"?

Reply:
Oh, to be clear, the other qualifiers *are* stripped because the underlying QualType loses them due to us using a type pointer.

Reply:
Right, so I guess I am wondering why the 'if' here, and not just 

`EnumUnderlying = TI->getType().getAtomicUnqualifiedType().getTypePtr();`.

without the `if` statement here.  Though I guess we have to make sure that `TI` is valid at least, so:

```
if (EnumUnderlying)
  EnumUnderlying = EnumUnderlying->getType().getAtomciUnqualifiedType().getTypePtr();
```

Side note: A little bit of a 'shadowing' problem happening here too.

Reply:
Ah, gross... but yeah, I guess so.  Can I have a 'fixme' on it?  Something like, "we COOOOULD just unconditionally strip this but it loses the TypeSourceInfo... someday figure out how to do that"